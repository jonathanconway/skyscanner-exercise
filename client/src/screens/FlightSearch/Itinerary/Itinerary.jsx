import React from 'react';
import PropTypes from 'prop-types';
import BpkText from 'bpk-component-text';
import BpkSmallLongArrowRight from 'bpk-component-icon/sm/long-arrow-right';
import BpkCard from 'bpk-component-card';
import BpkButton from 'bpk-component-button';
import BpkImage from 'bpk-component-image';

import { carrierFavIconHostAndPath } from '../../../config';
import { dateTime, misc } from '../../../utils';

import STYLES from './Itinerary.scss';

const c = className => STYLES[className] || 'UNKNOWN';

/**
 * Transforms an URL that points to a full-sized carrier image into
 * an URL that points to an icon-sized carrier image.
 * @param {string} imageUrl - Full-sized image URL
 * @return {string} - Icon-sized image URL
 */
const mapCarrierImageUrlToIconUrl = (imageUrl) => {
  return misc.replacePath(imageUrl, carrierFavIconHostAndPath);
};

const Itinerary = ({ id, legs, price }) => {
  return (
    <BpkCard key={id} className={c('Itinerary')}>
      {legs.map((leg, legIndex) => (
        <div key={`leg-${legIndex}`}>
          {leg.segments.map(({
            id,
            carrier,
            departure,
            arrival,
            duration
          }) => (
            <div key={id} className={c('Itinerary__segment')}>
              <BpkImage
                className={c('Itinerary__segment-carrier-icon')}
                altText={carrier.name}
                width={32}
                height={32}
                src={mapCarrierImageUrlToIconUrl(carrier.imageUrl)}
              />

              <div className={c('Itinerary__segment-sides')}>
                <div className={c('Itinerary__segment-sides-side')}>
                  <BpkText tag="span">
                    {dateTime.formatTime(departure.dateTime)}
                  </BpkText>
                  <BpkText tag="span" className={c('Itinerary__segment-sides-side-code')}>
                    {departure.airportCode}
                  </BpkText>
                </div>

                <b className={c('Itinerary__segment-sides-divider')}>
                  <BpkSmallLongArrowRight />
                </b>

                <div className={c('Itinerary__segment-sides-side')}>
                  <BpkText tag="span">
                    {dateTime.formatTime(arrival.dateTime)}
                  </BpkText>
                  <BpkText tag="span" className={c('Itinerary__segment-sides-side-code')}>
                    {arrival.airportCode}
                  </BpkText>
                </div>
              </div>

              <div className={c('Itinerary__segment-details')}>
                <BpkText tag="span">
                  {dateTime.formatDuration(duration)}
                </BpkText>
                <BpkText tag="span" className={c('Itinerary__segment-details-type-direct')}>
                  Direct
                </BpkText>
              </div>
            </div>
          ))}
        </div>
      ))}

      <div className={c('Itinerary__details')}>
        <div>
          <BpkText tag="span" textStyle="xl">
            {price.currency.symbol}{price.amount}
          </BpkText>
          <BpkText tag="span" textStyle="base" className={c('Itinerary__details-source')}>
            {price.agent.name}
          </BpkText>
        </div>
        <div>
          <BpkButton large={true} className={c('Itinerary__details-select-button')}>
            Select
          </BpkButton>
        </div>
      </div>
    </BpkCard>
  );
};

Itinerary.propTypes = {
  id: PropTypes.string,
  leg: PropTypes.shape({
    segments: PropTypes.arrayOf(PropTypes.shape({
      carrier: PropTypes.shape({
        name: PropTypes.string,
        imageUrl: PropTypes.string
      }),
      departure: PropTypes.shape({
        dateTime: PropTypes.string,
        airportCode: PropTypes.string,
      }),
      arrival: PropTypes.shape({
        dateTime: PropTypes.string,
        airportCode: PropTypes.string,
      }),
      duration: PropTypes.number
    })),
    price: PropTypes.shape({
      currencySymbol: PropTypes.string,
      amount: PropTypes.number,
      agent: PropTypes.shape({
        name: PropTypes.string
      })
    })
  })
};

export default Itinerary;
