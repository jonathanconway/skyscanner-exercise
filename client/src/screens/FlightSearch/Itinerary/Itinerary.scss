@import '~bpk-mixins';

@import '../../../theme';

.Itinerary {
  &:not(:last-child) {
    margin-bottom: $bpk-spacing-sm;
  }

  &__segment {
    display: flex;
    flex-direction: row;
    margin-bottom: $bpk-spacing-sm;

    & > * {
      &:not(:last-child) {
        margin-right: $bpk-spacing-sm;
      }
    }

    &-carrier-icon {
      padding-top: 0.5rem;
      width: $bpk-icon-size-lg;
    }

    &-sides {
      display: flex;
      flex-direction: row;

      & > * {
        &:not(:last-child) {
          margin-right: $bpk-spacing-sm;
        }
      }

      &-side {
        display: flex;
        flex-direction: column;

        &-code {
          color: $bpk-color-sky-gray-tint-04;
        }
      }

      &-divider {
        display: flex;
        align-items: center;
        fill: $bpk-color-sky-gray-tint-04;
      }
    }

    &-details {
      display: flex;
      flex-direction: column;
      flex: 1;
      text-align: right;

      &-type-direct {
        color: $action-color-shade-4;
      }
    }
  }

  &__details {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    & > div {
      display: flex;
      flex-direction: column;
    }

    &-source {
      color: $bpk-color-sky-gray-tint-04;
    }

    &-select-button {
      @extend .action-button;

      display: inline-block;
      margin-top: auto;
    }
  }
}
