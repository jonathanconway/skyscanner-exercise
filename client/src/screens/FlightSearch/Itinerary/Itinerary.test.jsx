import React from 'react';
import ReactDOM from 'react-dom';
import ShallowRenderer from 'react-test-renderer/shallow';

import Itinerary from './Itinerary';

describe('Itinerary', () => {
  const props = {
    id: 0,
    leg: {
      segments: [
        {
          id: 's0',
          carrier: {
            name: 'CARRIER_NAME',
            imageUrl: 'imageUrl.png'
          },
          departure: {
            dateTime: '2019-10-22T16:20:00',
            airportCode: 'EDI'
          },
          arrival: {
            dateTime: '2019-10-23T16:20:00',
            airportCode: 'LGW'
          },
          duration: 90
        },
        {
          id: 's1',
          carrier: {
            name: 'CARRIER_NAME',
            imageUrl: 'imageUrl.png'
          },
          departure: {
            dateTime: '2019-10-24T16:20:00',
            airportCode: 'LGW'
          },
          arrival: {
            dateTime: '2019-10-25T16:20:00',
            airportCode: 'EDI'
          },
          duration: 120
        }
      ]
    },
    price: {
      currency: {
        code: 'GBP',
        symbol: 'Â£'
      },
      amount: 123,
      agent: {
        name: 'ABC'
      }
    }
  };

  it('should render without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <Itinerary
        {...props}
      />, div);
  });

  it('should render correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Itinerary
        {...props}
      />
    );

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
