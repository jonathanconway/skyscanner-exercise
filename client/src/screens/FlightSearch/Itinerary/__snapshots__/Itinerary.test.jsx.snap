// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Itinerary should render correctly 1`] = `
<BpkCard
  blank={false}
  className="Itinerary"
  href={null}
  padded={true}
>
  <div>
    <div
      className="Itinerary__segment"
    >
      <BpkImage
        altText="CARRIER_NAME"
        borderRadiusStyle="none"
        className="Itinerary__segment-carrier-icon"
        height={32}
        inView={true}
        loading={false}
        onLoad={null}
        src="https://logos.skyscnr.com/images/airlines/favicon/imageUrl.png"
        style={Object {}}
        suppressHydrationWarning={false}
        width={32}
      />
      <div
        className="Itinerary__segment-sides"
      >
        <div
          className="Itinerary__segment-sides-side"
        >
          <BpkText
            bold={false}
            className={null}
            tag="span"
            tagName="span"
            textStyle="base"
          >
            16:20
          </BpkText>
          <BpkText
            bold={false}
            className="Itinerary__segment-sides-side-code"
            tag="span"
            tagName="span"
            textStyle="base"
          >
            EDI
          </BpkText>
        </div>
        <b
          className="Itinerary__segment-sides-divider"
        >
          <_default />
        </b>
        <div
          className="Itinerary__segment-sides-side"
        >
          <BpkText
            bold={false}
            className={null}
            tag="span"
            tagName="span"
            textStyle="base"
          >
            16:20
          </BpkText>
          <BpkText
            bold={false}
            className="Itinerary__segment-sides-side-code"
            tag="span"
            tagName="span"
            textStyle="base"
          >
            LGW
          </BpkText>
        </div>
      </div>
      <div
        className="Itinerary__segment-details"
      >
        <BpkText
          bold={false}
          className={null}
          tag="span"
          tagName="span"
          textStyle="base"
        >
          1h 30
        </BpkText>
        <BpkText
          bold={false}
          className="Itinerary__segment-details-type-direct"
          tag="span"
          tagName="span"
          textStyle="base"
        >
          Direct
        </BpkText>
      </div>
    </div>
  </div>
  <div>
    <div
      className="Itinerary__segment"
    >
      <BpkImage
        altText="CARRIER_NAME"
        borderRadiusStyle="none"
        className="Itinerary__segment-carrier-icon"
        height={32}
        inView={true}
        loading={false}
        onLoad={null}
        src="https://logos.skyscnr.com/images/airlines/favicon/imageUrl.png"
        style={Object {}}
        suppressHydrationWarning={false}
        width={32}
      />
      <div
        className="Itinerary__segment-sides"
      >
        <div
          className="Itinerary__segment-sides-side"
        >
          <BpkText
            bold={false}
            className={null}
            tag="span"
            tagName="span"
            textStyle="base"
          >
            16:20
          </BpkText>
          <BpkText
            bold={false}
            className="Itinerary__segment-sides-side-code"
            tag="span"
            tagName="span"
            textStyle="base"
          >
            LGW
          </BpkText>
        </div>
        <b
          className="Itinerary__segment-sides-divider"
        >
          <_default />
        </b>
        <div
          className="Itinerary__segment-sides-side"
        >
          <BpkText
            bold={false}
            className={null}
            tag="span"
            tagName="span"
            textStyle="base"
          >
            16:20
          </BpkText>
          <BpkText
            bold={false}
            className="Itinerary__segment-sides-side-code"
            tag="span"
            tagName="span"
            textStyle="base"
          >
            EDI
          </BpkText>
        </div>
      </div>
      <div
        className="Itinerary__segment-details"
      >
        <BpkText
          bold={false}
          className={null}
          tag="span"
          tagName="span"
          textStyle="base"
        >
          2h
        </BpkText>
        <BpkText
          bold={false}
          className="Itinerary__segment-details-type-direct"
          tag="span"
          tagName="span"
          textStyle="base"
        >
          Direct
        </BpkText>
      </div>
    </div>
  </div>
  <div
    className="Itinerary__details"
  >
    <div>
      <BpkText
        bold={false}
        className={null}
        tag="span"
        tagName="span"
        textStyle="xl"
      >
        Â£
        123.00
      </BpkText>
      <BpkText
        bold={false}
        className="Itinerary__details-source"
        tag="span"
        tagName="span"
        textStyle="base"
      >
        ABC
      </BpkText>
    </div>
    <div>
      <BpkButton
        blank={false}
        className="Itinerary__details-select-button"
        destructive={false}
        disabled={false}
        featured={false}
        href={null}
        iconOnly={false}
        large={true}
        link={false}
        onClick={null}
        outline={false}
        rel={null}
        secondary={false}
        submit={false}
      >
        Select
      </BpkButton>
    </div>
  </div>
</BpkCard>
`;
